packages/flutter/lib/src/widgets/scrollbar.dart

```diff
@@ -0,0 +1,240 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:math' as math;
+
+import 'package:flutter/animation.dart';
+import 'package:flutter/foundation.dart';
+import 'package:flutter/rendering.dart';
+import 'package:flutter/scheduler.dart';
+
+import 'scroll_metrics.dart';
+
+const double _kMinThumbExtent = 18.0;
+const Duration _kDefaultThumbFadeDuration = const Duration(milliseconds: 300);
+const Duration _kDefaultTimeToFade = const Duration(milliseconds: 600);
+
+/// A [CustomPainter] for painting scrollbars.
+///
+/// See also:
+///
+///  * [Scrollbar] for a widget showing a scrollbar around a [Scrollable] in the
+///    Material Design style.
+///  * [CupertinoScrollbar] for a widget showing a scrollbar around a
+///    [Scrollable] in the iOS style.
+class ScrollbarPainter extends ChangeNotifier implements CustomPainter {
+  /// Creates a scrollbar with customizations given by construction arguments.
+  ScrollbarPainter({
+    @required TickerProvider vsync,
+    @required this.thickness,
+    @required this.crossAxisMargin,
+    this.mainAxisMargin: 0.0,
+    this.radius,
+    this.minLength: _kMinThumbExtent,
+    this.timeToFadeout: _kDefaultTimeToFade,
+    this.fadeoutDuration: _kDefaultThumbFadeDuration,
+  })
+      : assert(vsync != null),
+        assert(thickness != null),
+        assert(mainAxisMargin != null),
+        assert(minLength != null),
+        assert(timeToFadeout != null),
+        assert(fadeoutDuration != null) {
+    _fadeController = new AnimationController(duration: fadeoutDuration, vsync: vsync);
+    _opacity = new CurvedAnimation(parent: _fadeController, curve: Curves.fastOutSlowIn)
+      ..addListener(notifyListeners);
+  }
+
+  /// Thickness of the scrollbar in its cross-axis in pixels. Mustn't be null.
+  double thickness;
```

Comment:
It doesn't make any sense for a CustomPainter to have externally mutable state.

Reply:
Maybe that's too strong a statement. Let me say instead, it's very much contrary to how CustomPainter was intended to work.

Reply:
Seems like the existing custom painter in material already mutates the painter (when the inherited theme changes). I didn't see any class docs on CustomPainter on its desired immutability.